network/router_test.go

```diff
@@ -0,0 +1,365 @@
+	}
+}
+
+// Test the automatic connection upon request
+func TestRouterAutoConnectionTCP(t *testing.T) {
+	testRouterAutoConnection(t, NewTestRouterTCP)
+}
+func TestRouterAutoConnectionLocal(t *testing.T) {
+	testRouterAutoConnection(t, NewTestRouterLocal)
+}
+
+func testRouterAutoConnection(t *testing.T, fac routerFactory) {
+	//	log.TestOutput(true, 5)
+	fmt.Println("Before fac()")
+	h1, err := fac(2007)
+	if err != nil {
+		t.Fatal(err)
+	}
+	fmt.Println("after fac()")
+	err = h1.Send(&ServerIdentity{Address: NewLocalAddress("127.1.2.3:2890")}, &SimpleMessage{12})
+	if err == nil {
+		t.Fatal("Should not be able to send")
+	}
+	fmt.Println("after send() #1")
+	h2, err := fac(2008)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = h1.Send(h2.id, nil)
+	if err == nil {
+		t.Fatal("Should not be able to send")
+	}
+
+	go h2.Start()
+
+	proc := newSimpleMessageProc(t)
+	h2.RegisterProcessor(proc, SimpleMessageType)
+	h1.RegisterProcessor(proc, SimpleMessageType)
+	defer func() {
+		assert.Nil(t, h1.Stop())
+		assert.Nil(t, h2.Stop())
+	}()
+
+	err = h1.Send(h2.id, &SimpleMessage{12})
+	if err != nil {
+		t.Fatal("Couldn't send message:", err)
+	}
+
+	// Receive the message
+	msg := <-proc.relay
+	if msg.I != 12 {
+		t.Fatal("Simple message got distorted")
+	}
+
+	h12 := h1.connection(h2.id.ID)
+	h21 := h2.connection(h1.id.ID)
+	if h12 == nil {
+		t.Error("h1 has no connection to h2")
+	} else if h21 == nil {
+		t.Error("h2 has no connection to h1")
+	}
+
+	assert.Nil(t, h21.Close())
+	if err := h2.Stop(); err != nil {
+		t.Fatal("Should be able to stop h2")
+	}
+	err = h1.Send(h2.id, &SimpleMessage{12})
+	if err == nil {
+		t.Fatal("Should not be able to send stuff !!")
+	}
+}
+
+// Test connection of multiple Hosts and sending messages back and forth
+// also tests for the counterIO interface that it works well
+func TestRouterMessaging(t *testing.T) {
+	h1, err1 := NewTestRouterTCP(2009)
+	h2, err2 := NewTestRouterTCP(2010)
+	if err1 != nil || err2 != nil {
+		t.Fatal("Could not setup hosts")
+	}
+
+	go h1.Start()
+	go h2.Start()
+
+	defer func() {
+		h1.Stop()
+		h2.Stop()
+		time.Sleep(250 * time.Millisecond)
+	}()
+
+	proc := &simpleMessageProc{t, make(chan SimpleMessage)}
+	h1.RegisterProcessor(proc, SimpleMessageType)
+	h2.RegisterProcessor(proc, SimpleMessageType)
+
+	msgSimple := &SimpleMessage{3}
+	err := h1.Send(h2.id, msgSimple)
+	if err != nil {
+		t.Fatal("Couldn't send from h2 -> h1:", err)
+	}
+	decoded := <-proc.relay
+	if decoded.I != 3 {
+		t.Fatal("Received message from h2 -> h1 is wrong")
+	}
+
+	// make sure the connection is registered in host1 (because it's launched in
+	// a go routine). Since we try to avoid random timeout, let's send a msg
+	// from host2 -> host1.
+	assert.Nil(t, h2.Send(h1.id, msgSimple))
+	decoded = <-proc.relay
+	assert.Equal(t, 3, decoded.I)
+
+	written := h1.Tx()
+	read := h2.Rx()
+	if written == 0 || read == 0 || written != read {
+		t.Logf("Tx = %d, Rx = %d", written, read)
```

Comment:
Why ? It's something that is not supposed to happen, why not using the testing output which is exactly made for that ?